[MIG][16.0] product_analytic_sale

rousseldenis opened the pull request.

alexis-via committed:
[ADD] product_analytic_sale

The readme gives an example of a scenario where this module product_analytic_sale is required.

clementmbr committed:
[FIX] check if SO line is not a type note (no product)

Commit:
[UPD] Update product_analytic_sale.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

bizzappdev committed:
[MIG][15.0]product_analytic_sale:Migration to 15.0

rousseldenis committed:
[IMP] product_analytic_sale: pre-commit stuff

rousseldenis committed:
[MIG][16.0] product_analytic_sale

rousseldenis commented:
@alexis-via @bizzappdev

rousseldenis commented:
@sbidoul

bizzappdev commented on product_analytic_sale/models/sale_order_line.py:
```diff
@@ -0,0 +1,46 @@
+        Get analytic distribution from product expense analytic account
+        If no account is set, call super with those records
+        """
+        lines_without_analytic_ids = set()
+        for line in self:
+            if line.product_id:
+                ana_accounts = (
+                    line.product_id.product_tmpl_id._get_product_analytic_accounts()
+                )
+                ana_account = ana_accounts["income"]
+                if ana_account:
+                    line.analytic_distribution = {ana_account.id: 100}
+                else:
+                    lines_without_analytic_ids.add(line.id)
+        return super(
+            SaleOrderLine, self.browse(lines_without_analytic_ids)
```
```suggestion
        lines_without_analytic_ids = self.env["sale.order.line"]
        for line in self:
            if line.product_id:
                ana_accounts = (
                    line.product_id.product_tmpl_id._get_product_analytic_accounts()
                )
                ana_account = ana_accounts["income"]
                if ana_account:
                    line.analytic_distribution = {ana_account.id: 100}
                else:
                    lines_without_analytic_ids |= line
            else:
                lines_without_analytic_ids |= line
        return super(
            SaleOrderLine, lines_without_analytic_ids
```

bizzappdev replied:
IMO, this will save browsing again already browsed

rousseldenis replied:
Indeed, before v16, union was quite not performant. It operates quite flawlessly now, even if last browse does not affect performances as ids already in cache.

rousseldenis replied:
@bizzappdev Code updated

sbidoul commented on product_analytic_sale/models/sale_order_line.py:
```diff
@@ -0,0 +1,46 @@
+                    lines_without_analytic_ids |= line
+        return super(
+            SaleOrderLine, lines_without_analytic_ids
+        )._compute_analytic_distribution()
```
In #576 (product_analytic_purchase), `create` is also overridden. Why is it different here?

rousseldenis replied:
Inherited from old version of the module as the distribution was not computed. Not needed. To be changed there.

bizzappdev commented on product_analytic_sale/models/sale_order_line.py:
```diff
@@ -0,0 +1,46 @@
+        """
+        lines_without_analytic_ids = self.browse()
+        for line in self:
+            if line.product_id:
```
Might be I am missing something but don't you there should be else condition for this also so recodrs will be covered in super

rousseldenis replied:
Totally right

sbidoul commented on product_analytic_sale/models/sale_order_line.py:
```diff
@@ -0,0 +1,46 @@
+            )
+            if ana_account:
+                self.analytic_distribution = {ana_account.id: 100}
+        return vals
```
I have tested with standard Odoo:
- if there is an analytic account on product or category, it is added on the invoice, by the `product_analytic` module
- if an analytic distribution is set on the SO line, it has priority over product analytic

So I think this `_prepare_invoice_line` override is not necessary anymore.

rousseldenis committed:
[IMP] product_analytic_sale: Rely on product_analytic module for invoices + distribution model

Don't override anymore the _prepare_invoice_line() function as product_analytic module does
the same job.

If no analytic account or no product is set, let the super() happens.

rousseldenis commented:
@sbidoul @bizzappdev Last check, this one is :green_apple:  ?

sbidoul reviewed (commented):
LGTM. I'll wait for our integration tests for giving the approval here.

sbidoul commented on product_analytic_sale/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,10 @@
+
+This module is a glue module between the modules **product_analytic** and **sale**.
+
+The module **product_analytic** inherits the *create()* method of account.move.line to add the Analytic Account depending on the Product. But it is not enough in some specific scenarios: for example, if you have *Analytic Policy* set to *Required* on Revenue accounts and you have a sale tax configured with a revenue account and with the option *Include in Analytic Cost* enabled, then Odoo will no ...
```
@rousseldenis this paragraph is obsolete?

rousseldenis replied:
Mmmmh, I will add the test :sweat_smile:

rousseldenis replied:
@sbidoul In fact, this is covered by : https://github.com/OCA/account-analytic/pull/621/commits/5b203de8f6636c8bf7fafb28639955098c8fc473

sbidoul commented on product_analytic_sale/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,10 @@
+from the one stored either on product category either on product level
+(see product_analytic module).
+
+Developers
+==========
+
```
```suggestion
```

rousseldenis replied:
Why ? This is a technical section for developers.

sbidoul replied:
Because we must not add top level section titles in README fragments as this breaks the rendered document.

rousseldenis replied:
aaaah ok. Is this better ?

sbidoul replied:
Nope. If you really want a subsection, then use `~`, but really there is no reason to create a subsection just for one line.

PauMAVA commented:
@rousseldenis @sbidoul will this be merged soon? I would be interested in using it in production. Thanks!

rousseldenis commented:
> @rousseldenis @sbidoul will this be merged soon? I would be interested in using it in production. Thanks!

Ususally, some reviews are required in OCA to be merged. Don't hesitate to do it :smiley:

aritzolea reviewed (approved):
LGTM (technical review)

sbidoul commented:
@rousseldenis if I'm not mistaken you still need to handle my comments above?

rousseldenis commented:
@sbidoul This is ready

sbidoul commented on product_analytic_sale/models/sale_order_line.py:
```diff
@@ -0,0 +1,29 @@
+
+    def _compute_analytic_distribution(self):
+        """
+        Get analytic distribution from product expense analytic account
```
Why do we mention `expense` here?

houssine78 replied:
@rousseldenis did you get time to go ahead with the comment of Stéphane ?

houssine78 reviewed (approved):
LGTM

dreispt commented:
/ocabot migration product_analytic_sale
/ocabot merge nobump
